### DISCUSSION TITLE
boardfunctions.py, epaper.py, gamemanager?

### DISCUSSION THREAD

**Discussion by EdNekebno** [EdNekebno/DGTCentaurMods#47] [Ideas]
We have the project board. But I thought it useful describe my plans/thoughts - in case there are any ideas/suggestions/modifications or it just won't work :)

I have recreated better epaper screen functions in epaper.py . Most scripts I've migrated to this now. I'll leave the old screen functions in boardfunctions.py for legacy and because at least on things can't (my wifi setup password thingy).

At the end of boardfunctions.py I have added a method to subscribe to events from the board. This means if you write a script then it can be passed keypresses and data when a piece is lifted or placed back on the board.

I'll do a general tidy up of boardfunctions.py and sort it into sections with types of functions - at the moment it is quite messy.

eboard.py is pretty much done. It sits on the "In Progress" lane in the projects tab because there's still some testing. But I'm pretty happy with the fact it emulates a bluetooth e-board and that the piece movement is detection seems pretty solid.

lichess.py - I am told is good. This sits on the "In Progress" lane because it has contributed code and I haven't tested it much myself.

My plan is to use the piece movement detection from eboard.py and create a game manager script. This will allow a script to subscribe to more than just piece up/down events, but will manage a whole game and pass through things like information that a new game has started, the actual moves, and handle any mistakes. Let's assume for now that's gamemanager.py

Then most actual game playing scripts will be based off those events. It'd then be a plan to use that to create a script that can call any UCI engine we want, with any settings we want. This means, for example, that the user can play stockfish or others directly.

I'd then update the lichess and eboard scripts to use this new mechanism. The advantage here is that at this point all of the scripts are using the same base mechanism. A database, like the suggest SQLite, becomes really useful here. As every game can be databased for playback, analysis, or PGN export (from the cool web interface. although I do like the idea of getting some basic game stats on the screen. There's also the idea that you could play back a game on the board and the epaper can point out where your move was wrong and better lines).

With a UCI script done, it's not that much work to do the adaptive play thing and essentially make an alternative to the centaur process. Just because we can and it gives an option to have something that is independent of the original software.

Of course, in there there might come some other board emulations. I think emulating other boards to make it work with a lot of different things/apps is a good direction to go because the apps, for example, automatically offer similar features - e.g. analysis, PGN (just not on the board itself).

Oh and I have one idea that I'm going to keep mysteriously secret in case it doesn't work. But if you haven't noticed by now I get distracted with side projects a lot :)

**Comment by wormstein** [comment]
I think we posted at the same time :)) I was asking in the other topic if you think the epaper.py would be better to sit in display section?

The progress you did is insane!! Good job!

**Comment by EdNekebno** [comment]
epaper is now in the display, I also started on gamemanager.py and added an examples folder for coding examples so I could give an example for it. (examples are intended not to go on the menu but to serve as a starting point if somebody wants to code something).

**Reply by wormstein** [reply]
That's good to have some examples on new features. Took me some time the other day to figure out how follow your code. Having some comments as well helps.

I'll make sure our releases will skip to include these files...

**Comment by wormstein** [comment]
Looking over game manager and what you try to achieve, isn't this easier with python chess.pgn?

And you just need to store game information in db, and moves in a file that you reference in db.

Chess.pgn also provide an easier way to parse the moves and reconstruct the game.

Just an idea, hope i got it right your goal.

Cheers

**Comment by EdNekebno** [comment]
If you mean the PGN bit, I agree but I mean without the 'file' because you'd have lots of files.

Basically gamemanager manages a chess game. Then other scripts just plug into that. So the engine playing scripts, for example, do very little but get moves from the engine. This makes gamemanager somewhat central. So it records everything to a database. Both the games, and the moves (saves having lots of files). Basically it keeps you data for every game you've played - which can be useful for things like tracking progress, frequent mistakes, etc. Theoretically I could find a game in the database that I played last week, in eboard mode, even though I may have played 20 games against an again since then.

In terms of producing PGNs. It's probably easiest from the web interface. But it would work pretty much like that. As in you pick your game from the game table and loop through the moves in the db (probably with chess.pgn because why re-invent the wheel). If we ever do analysis on the board itself as an option too - it'd be a similar sort of thing - loop through but just send the moves to stockfish for analysis.

**Comment by wormstein** [comment]
Makes sense what your goal is.

I am thinking of a way to keep this file out of the repo but at the  moment we just have to avoid commiting it., Only when structure changes.

**Comment by wormstein** [comment]
Morning,
about for gamemanager for centaur, I have one suggestion

Gamemaneger works fine when playing Lichess, eboard and other ones integrated in DGTCentaurMods, but once the centaur software is started, the Mods are turned off. You mentioned you look to do a check fot centaur process. I'd say that might be too hard on the board as you need to frequently check so that you don't miss the start of the game.

I would suggest for the centaur software to have a separate python script to handle writing to DB, which we can fire up as a service in systemd pre starting centaur.service. Something called centaur-helper.service.

Or, If you manage to modify stockfish to write to DB that's good. Then you can write a new game event to DB just before centaur software is started.

**Comment by EdNekebno** [comment]
Yes, this is what I meant I just didn't explain it very well.

I have modified my stockfish so that it does the following:

1. If the uci command is "position startpos" it starts a new game in the game table in the database
2. If the uci command is "position startpos moves ... ..." it writes the move to the gameMove table in the database

Here the problem is if something calls stockfish with uci through gamemanager - e.g. my stockfish script - then it would get two writes (one from gamemanager and one from stockfish).

So before doing 1 and 2 I have coded stockfish to check if the centaur process is running. If it is running then it knows that it should do it. If it is not running then it knows not to do it as gamemanager will take care of it.

I'm testing the modified stockfish at the moment. It looks promising.

**Reply by wormstein** [reply]
Sounds very good how you handle this. I guess I need to grab and compile SF from you.

**Reply by EdNekebno** [reply]
I think the license is that I need to provide code changes. I'll fork your fork of the original and update it, just to make sure we're playing by all the rules.

**Reply by EdNekebno** [reply]
https://github.com/EdNekebno/Stockfish

In uci.cpp I refer to the location of the centaur.db file twice. You may need to change it if it is a different location. Then compile. Also as we discussed the tables need to be in the database too. So either have the latest db upload or run something else based off gamemanager to create them.
